File: deadlock/deadlock.cpp
#include <iostream>
#include <mutex>
#include <thread>

using namespace std;

static mutex m1, m2;

void bob()
{
	lock_guard<mutex> g1(m1);
	lock_guard<mutex> g2(m2);
	cout << "Hi, it's Bob" << endl;
}

void alice()
{
	lock_guard<mutex> g1(m2);
	lock_guard<mutex> g2(m1);
	cout << "Hi, it's Alice" << endl;
}

int main()
{
	thread t_bob(bob);
	thread t_alice(alice);

	t_bob.join();
	t_alice.join();
	return 0;
}

File: deadlock/Deadlock.java
public class Deadlock {
    public static void main(String[] args) {
        Object mutex1 = new Object();
        Object mutex2 = new Object();

        Runnable bob = () -> {
            for (int i = 0; i < 1000; i++)
                synchronized(mutex1) {
                    synchronized(mutex2) {
                        System.out.println("Hi, it's Bob " + i);
                    }
                }
        };
        Runnable alice = () -> {
            for (int i = 0; i < 1000; i++)
                synchronized(mutex2) {
                    synchronized(mutex1) {
                        System.out.println("Hi, it's Alice " + i);
                    }
                }
        };

        Thread at = new Thread(alice);
        Thread bt = new Thread(bob);

        bt.start();
        at.start();

        try {
            at.join();
            bt.join();
        } catch(InterruptedException e) {
            System.err.println("Interrupted: " + e);
        }
    }
}
